src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs

```diff
@@ -30,36 +30,127 @@ public ChunkedEncodingReadStream(HttpConnection connection) : base(connection)
             {
             }
 
+            /// <summary>Reads chunked data from the connection's buffer into the destination.</summary>
+            /// <param name="destination">The destination into which bytes will be copied.</param>
+            /// <returns>The number of bytes written to <paramref name="destination"/>.</returns>
+            private int ReadChunksFromBuffer(Span<byte> destination)
+            {
+                Debug.Assert(_chunkBytesRemaining == 0, $"{nameof(ReadChunksFromBuffer)} should only be used at the beginning of a chunk.");
+
+                int totalBytesConsumed = 0;
+                while (destination.Length > 0)
+                {
+                    ReadOnlySpan<byte> connectionBuffer = _connection.RemainingBuffer;
+                    if (connectionBuffer.Length < 2 || // at a minimum we need one number and a \n, though most likely at least \r\n
+                        !Utf8Parser.TryParse(connectionBuffer, out ulong chunkSize, out int bytesConsumed, 'X') || // Needs to start with a chunk size
```

Comment:
Why not look for \n first?  Seems like we could end up parsing the chunkSize multiple times here.

Reply:
Also, seems like you could have a function like ReadOnlySpan<byte> TryReadLineFromBuffer.  That seems useful in two places here, possibly others outside of this...

Reply:
> Why not look for \n first? Seems like we could end up parsing the chunkSize multiple times here.

We could end up parsing multiple times if either a) the data is malformed or b) the line isn't fully buffered.  I don't think we care about (a), and I expect it'll be relatively rare for the line to not be fully buffered, as in general (obviously not always) I expect it'd all be flushed by the server and received in the same packet.  And if we assume the success case, then it's (slightly) more efficient to parse the variable-length value and then only search the remainder of the line for the \n.

Reply:
> Also, seems like you could have a function like ReadOnlySpan TryReadLineFromBuffer. That seems useful in two places here, possibly others outside of this

I actually started with that as I agree that seemed most reusable.  But then I ran into some issue which for the life of me I can't remember right now and backed off of it.  I'll try again.